metagraph/core/resolver.py

```diff
@@ -495,6 +537,24 @@ def _normalize_abstract_algorithm_signature(self, abst_algo: AbstractAlgorithm):
                 p = p.replace(annotation=pmod)
                 any_changed = True
             params_modified.append(p)
+
+        self._normalize_abstract_algorithm_return_type(
+            abst_algo, params_modified, any_changed
+        )
+
+        return abst_algo
+
+    def _normalize_abstract_algorithm_return_type(
```

Comment:
I wouldn't break this piece of logic out. Yes, it makes two shorter functions, but the call to `_normalize_abstract_algorithm_return_type`, its definition and signature, plus comments and redefinition of abs_sig makes the code somewhat less readable.

Additionally, having the return type normalization update the params feels strange. It means you can't reorder where in the calling function things happen.

A harder, but better refactor would be to strip out all the `change`, `any_changed`, `p.replace`, and `abs_sig.replace` logic into a separate class. Make it hold a function signature. Pass it to `_check_abstract_type` and make modification calls on the object. Let all the logic live there.

This logic is repeated for the normalization of concrete algorithm signatures, so this would definitely help clean things up.

Reply:
https://github.com/metagraph-dev/metagraph/pull/169/commits/dceed9c8ff1c995c34a4a679f60597dd9b1ab03d implements this suggestion.